File: src/Apps/ArcaneKingdom/Unity/Assets/_Project/Scripts/Domain/Battle/ElementMatchup.cs
#nullable enable
using ArcaneKingdom.Domain.Cards;

namespace ArcaneKingdom.Domain.Battle
{
    /// <summary>
    /// Element-Effektivitäts-Matrix für das Doppel-Dreieck-System (Designplan v4 Kap. 3).
    ///
    /// Physisches Dreieck:
    ///   Feuer  → stark gegen Natur
    ///   Natur  → stark gegen Wasser
    ///   Wasser → stark gegen Feuer
    ///
    /// Magisches Dreieck:
    ///   Dunkel → stark gegen Erde   (Verderbnis zersetzt die Erde)
    ///   Erde   → stark gegen Licht  (Verschüttung, Finsternis)
    ///   Licht  → stark gegen Dunkel (Lichtschwert vertreibt die Dunkelheit)
    ///
    /// Stark = +10% Schaden + Element-Spezialeffekt (Verbrennung, Einfrierung, Steinpanzer etc.).
    /// Designplan v4 Kap. 3.3 kennt NUR den +10%-Bonus für das überlegene Element — KEINEN
    /// Schwäche-Malus. Alle nicht-starken Paarungen (inkl. Rückrichtung) sind daher neutral (1.0x).
    /// </summary>
    public static class ElementMatchup
    {
        /// <summary>Neutraler Multiplikator (1.0x) — auch für die "schwache" Rückrichtung (v4 hat keinen Malus).</summary>
        public const float NeutralMultiplier = 1.0f;

        /// <summary>Stark gegen — +10% Schaden (Designplan v4 Kap. 3.3).</summary>
        public const float StrongMultiplier = 1.10f;

        public static float GetMultiplier(Element attacker, Element defender)
        {
            return (attacker, defender) switch
            {
                // Physisches Dreieck: Feuer → Natur → Wasser → Feuer (nur +10%, kein Rück-Malus)
                (Element.Feuer,  Element.Natur)  => StrongMultiplier,
                (Element.Natur,  Element.Wasser) => StrongMultiplier,
                (Element.Wasser, Element.Feuer)  => StrongMultiplier,

                // Magisches Dreieck: Licht → Dunkel → Erde → Licht
                (Element.Licht,  Element.Dunkel) => StrongMultiplier,
                (Element.Dunkel, Element.Erde)   => StrongMultiplier,
                (Element.Erde,   Element.Licht)  => StrongMultiplier,

                // Alle anderen Paarungen (Schwäche-Rückrichtung + Cross-Dreieck): neutral
                _ => NeutralMultiplier
            };
        }

        /// <summary>
        /// Liefert true, wenn das angreifende Element zum physischen Dreieck (Feuer/Wasser/Natur) gehört.
        /// </summary>
        public static bool IsPhysical(Element element) =>
            element is Element.Feuer or Element.Wasser or Element.Natur;

        /// <summary>
        /// Liefert true, wenn das Element zum magischen Dreieck (Licht/Dunkel/Erde) gehört.
        /// </summary>
        public static bool IsMagical(Element element) =>
            element is Element.Licht or Element.Dunkel or Element.Erde;
    }
}

File: src/Apps/ArcaneKingdom/Server/CloudFunctions/src/battle/engine/elementMatchup.ts
// Element-Effektivitaets-Matrix — 1:1-Portierung von
// Unity/Assets/_Project/Scripts/Domain/Battle/ElementMatchup.cs.
//
// Doppel-Dreieck (Designplan v4 Kap. 3):
//   Physisch: Feuer -> Natur -> Wasser -> Feuer
//   Magisch:  Licht -> Dunkel -> Erde -> Licht
//   Stark = 1.10x, Schwach = 0.90x, Cross-Dreieck = 1.00x (neutral).
//
// Determinismus-Hinweis: Die Multiplier sind exakt dieselben float-Konstanten wie in C#
// (1.0 / 1.10 / 0.90). Sie gehen anschliessend durch `(int)(damage * multiplier)` (Truncation).

import { Element } from "./types";

export const NEUTRAL_MULTIPLIER = 1.0;
export const STRONG_MULTIPLIER = 1.1;
export const WEAK_MULTIPLIER = 0.9;

/** Liefert den Schadens-Multiplikator des Angreifer-Elements gegen das Verteidiger-Element. */
export function getMultiplier(attacker: Element, defender: Element): number {
  // Physisches Dreieck: Feuer -> Natur -> Wasser -> Feuer
  if (attacker === Element.Feuer && defender === Element.Natur) return STRONG_MULTIPLIER;
  if (attacker === Element.Natur && defender === Element.Wasser) return STRONG_MULTIPLIER;
  if (attacker === Element.Wasser && defender === Element.Feuer) return STRONG_MULTIPLIER;
  if (attacker === Element.Natur && defender === Element.Feuer) return WEAK_MULTIPLIER;
  if (attacker === Element.Wasser && defender === Element.Natur) return WEAK_MULTIPLIER;
  if (attacker === Element.Feuer && defender === Element.Wasser) return WEAK_MULTIPLIER;

  // Magisches Dreieck: Licht -> Dunkel -> Erde -> Licht
  if (attacker === Element.Licht && defender === Element.Dunkel) return STRONG_MULTIPLIER;
  if (attacker === Element.Dunkel && defender === Element.Erde) return STRONG_MULTIPLIER;
  if (attacker === Element.Erde && defender === Element.Licht) return STRONG_MULTIPLIER;
  if (attacker === Element.Dunkel && defender === Element.Licht) return WEAK_MULTIPLIER;
  if (attacker === Element.Erde && defender === Element.Dunkel) return WEAK_MULTIPLIER;
  if (attacker === Element.Licht && defender === Element.Erde) return WEAK_MULTIPLIER;

  // Verschiedene Dreiecke: neutral
  return NEUTRAL_MULTIPLIER;
}

/** True, wenn das Element zum physischen Dreieck (Feuer/Wasser/Natur) gehoert. */
export function isPhysical(element: Element): boolean {
  return element === Element.Feuer || element === Element.Wasser || element === Element.Natur;
}

/** True, wenn das Element zum magischen Dreieck (Licht/Dunkel/Erde) gehoert. */
export function isMagical(element: Element): boolean {
  return element === Element.Licht || element === Element.Dunkel || element === Element.Erde;
}
